man/dlopen-notes.1
.TH DLOPEN\-NOTES 1 "May 2024"
.SH NAME
dlopen\-notes \- parses dlopen elf metadata
.SH SYNOPSIS
.B dlopen\-notes
.RI [ OPTION ...]
.SH DESCRIPTION
ELF binaries store link-time dependencies in their headers, which can be parsed
by various tools. There is no machine-readable metadata about dependencies
loaded at build time via
.BR \%dlopen (3)
available by default. The ELF Dlopen Metadata specification aims to fill this
gap, by defining a common format.
.PP
This tool allows parsing such a note, and printing out the result in various
formats.
.SH OPTIONS
.TP
.BI \-\-raw
Show the original JSON extracted from input files
.TP
.BI \-\-sonames
List all sonames and their priorities, one soname per line
.TP
.BI \-\-features= [FEATURE...]
Print the description of the specified features
.TP
.BR \-h ", " \-\-help
Show a short help message and exit.
.SH SEE ALSO
.ad l
.nh
.BR dh\-dlopenlibdeps (1),
.UR https://systemd.io/ELF_DLOPEN_METADATA/
.B Dlopen Metadata for ELF Files
.UE


debian/changelog
package-notes (0.12) unstable; urgency=medium

  [ Luca Boccassi ]
  * d/control: mark python3 dependency with 'any' as per multiarch hinter

  [ Zbigniew Jędrzejewski-Szmek ]
  * dlopen-notes: make --raw the default action
  * dlopen-notes: add short option args and forbid abbrevs
  * dlopen-notes: generate man page using argparse-manpage

  [ Luca Boccassi ]
  * manpage: check in git
  * makefile: install manpage too
  * debian: install dlopen-notes manpage

 -- Luca Boccassi <bluca@debian.org>  Wed, 22 May 2024 15:35:08 +0100

package-notes (0.11) unstable; urgency=medium

  * d/control: drop 'native' from build dependencies
  * d/control: add doc link for dlopen spec

 -- Luca Boccassi <bluca@debian.org>  Wed, 15 May 2024 18:22:46 +0100

package-notes (0.10) unstable; urgency=medium

  * Add Python script to parse dlopen ELF notes
  * Add Debian tool for dlopen-notes
  * d/control: bump Standards-Version to 4.7.0, no changes

 -- Luca Boccassi <bluca@debian.org>  Thu, 09 May 2024 10:17:51 +0100

package-notes (0.9) unstable; urgency=medium

  * Switch to --package-metadata linker flag via specs file

 -- Luca Boccassi <bluca@debian.org>  Sun, 27 Nov 2022 16:43:58 +0000

package-notes (0.8) unstable; urgency=medium

  * use LONG instead of BYTEs for binary fields

 -- Luca Boccassi <bluca@debian.org>  Mon, 04 Apr 2022 19:52:18 +0100

package-notes (0.7) unstable; urgency=medium

  * dh: derive distro name at build time
  * dh: pick debuginfod from DEBUGINFOD_URLS env var if available
  * dh: install makefile include to automatically set LDFLAGS

 -- Luca Boccassi <bluca@debian.org>  Mon, 07 Mar 2022 11:40:22 +0000

package-notes (0.6) unstable; urgency=medium

  * Mark package as Multi-Arch: foreign

 -- Luca Boccassi <bluca@debian.org>  Sun, 20 Feb 2022 22:16:03 +0000

package-notes (0.5) unstable; urgency=medium

  [ Zbigniew Jędrzejewski-Szmek ]
  * generator: add a hint how to use the output to the output
  * generator: allow disabling READONLY
  * rpm: skip (READONLY) with older linker versions
  * Use -Wl,-dT
  * generator: add help strings

  [ Luca Boccassi ]
  * Initial packaging (Closes: #1004357)
  * Add architecture field
  * Skip fields without value
  * Debian: use -Wl,dT

  [ Zbigniew Jędrzejewski-Szmek ]
  * generator: add description in help output
  * generator: add --version argument
  * generator: disallow abbreviated option names
  * generator: rename program to not have .py suffix
  * generator: fix two minor issues reported by pylint
  * debian: stop using dh-exec

  [ Mark Wielaard ]
  * generate-package-notes.py: Add --debuginfod argument

  [ Victor Westerhuis ]
  * Add Debian debuginfod server in dh_package_notes
  * Provide dh-sequence-package_notes
  * Fix Python installation
  * Add manpage for generate-package-notes

  [ Zbigniew Jędrzejewski-Szmek ]
  * rpm: rename srpm macros file
  * rpm: fix filename in macro
  * rpm: conditionalize all macros on %_package_note_file being defined
  * rpm: make _generate_package_note_file always recreate the file
  * rpm: use $RPM_PACKAGE_VERSION variable to refer to the package version
  * rpm: also voidify the macros if we're on a noarch build
  * rpm: use %{buildsubdir} in %_package_note_file if defined
  * rpm: allow unsetting %_package_note_readonly to drop the READONLY
    attribute
  * rpm: add %_package_note_linker and document everything (rhbz#2043178,
    rhbz#2043368)
  * rpm: disable notes with linkers other than bfd
  * rpm: disable notes when clang toolchain is used on arm
  * generator: do not write zeroes to the linker script

  [ Frantisek Sumsal ]
  * ci: update repos before installing packages

  [ Zbigniew Jędrzejewski-Szmek ]
  * rpm: use named arguments

  [ Luca Boccassi ]
  * python: add --cpe auto
  * shell: add --cpe auto
  * man: document new --cpe auto option
  * shell: add --root option
  * python: add --root option
  * man: document new --root option
  * python: parse files in gather_data rather than parse_args
  * tests: add resources and tests for --auto-cpe and --root
  * rpm: use --cpe auto
  * debian: bump debhelper-compat version to 13
  * README: note we have both python and shell implementations

 -- Luca Boccassi <bluca@debian.org>  Mon, 31 Jan 2022 23:43:02 +0000


dlopen-notes.py
#!/usr/bin/env python3
# SPDX-License-Identifier: CC0-1.0

"""\
Read .note.dlopen notes from ELF files and report the contents.
"""

import argparse
import enum
import functools
import json
import sys
from elftools.elf.elffile import ELFFile
from elftools.elf.sections import NoteSection

try:
    import rich
    print_json = rich.print_json
except ImportError:
    print_json = print

def listify(f):
    def wrap(*args, **kwargs):
        return list(f(*args, **kwargs))
    return functools.update_wrapper(wrap, f)

@listify
def read_dlopen_notes(filename):
    elffile = ELFFile(open(filename, 'rb'))

    for section in elffile.iter_sections():
        if not isinstance(section, NoteSection) or section.name != '.note.dlopen':
            continue

        for note in section.iter_notes():
            if note['n_type'] != 0x407c0c0a or note['n_name'] != 'FDO':
                continue
            note_desc = note['n_desc']

            try:
                # On older Python versions (e.g.: Ubuntu 22.04) we get a string, on
                # newer versions a bytestring
                if not isinstance(note_desc, str):
                    text = note_desc.decode('utf-8').rstrip('\0')
                else:
                    text = note_desc.rstrip('\0')
            except UnicodeDecodeError as e:
                raise ValueError(f'{filename}: Invalid UTF-8 in .note.dlopen n_desc') from e

            try:
                j = json.loads(text)
            except json.JSONDecodeError as e:
                raise ValueError(f'{filename}: Invalid JSON in .note.dlopen note_desc') from e

            if not isinstance(j, list):
                print(f'{filename}: ignoring .note.dlopen n_desc with JSON that is not a list',
                      file=sys.stderr)
                continue

            yield from j

def dictify(f):
    def wrap(*args, **kwargs):
        return dict(f(*args, **kwargs))
    return functools.update_wrapper(wrap, f)

@dictify
def group_by_soname(notes):
    for note in notes:
        for element in note:
            priority = element.get('priority', 'recommended')
            for soname in element['soname']:
                yield soname, priority

class Priority(enum.Enum):
    suggested   = 1
    recommended = 2
    required    = 3

    def __lt__(self, other):
        return self.value < other.value

def group_by_feature(filenames, notes):
    features = {}

    # We expect each note to be in the format:
    # [
    #   {
    #     "feature": "...",
    #     "description": "...",
    #     "priority": "required"|"recommended"|"suggested",
    #     "soname": ["..."],
    #   }
    # ]
    for filename, note_group in zip(filenames, notes):
        for note in note_group:
            prio = Priority[note.get('priority', 'recommened')]
            feature_name = note['feature']

            try:
                feature = features[feature_name]
            except KeyError:
                # Create new
                feature = features[feature_name] = {
                    'description': note.get('description', ''),
                    'sonames': { soname:prio for soname in note['soname'] },
                }
            else:
                # Merge
                if feature['description'] != note.get('description', ''):
                    print(f"{filename}: feature {note['feature']!r} found with different description, ignoring",
                          file=sys.stderr)

                for soname in note['soname']:
                    highest = max(feature['sonames'].get(soname, Priority.suggested),
                                  prio)
                    feature['sonames'][soname] = highest

    return features

def make_parser():
    p = argparse.ArgumentParser(
        description=__doc__,
        allow_abbrev=False,
        add_help=False,
        epilog='If no option is specifed, --raw is the default.',
    )
    p.add_argument('-r', '--raw',
                   action='store_true',
                   help='Show the original JSON extracted from input files')
    p.add_argument('-s', '--sonames',
                   action='store_true',
                   help='List all sonames and their priorities, one soname per line')
    p.add_argument('-f', '--features',
                   nargs='?',
                   const=[],
                   type=lambda s: s.split(','),
                   action='extend',
                   metavar='FEATURE1,FEATURE2',
                   help='Describe features, can be specified multiple times')
    p.add_argument('filenames',
                   nargs='+',
                   metavar='filename',
                   help='Library file to extract notes from')
    p.add_argument('-h', '--help',
                   action='help',
                   help='Show this help message and exit')
    return p

def parse_args():
    args = make_parser().parse_args()

    if not args.raw and args.features is None and not args.sonames:
        # Make --raw the default if no action is specified.
        args.raw = True

    return args

if __name__ == '__main__':
    args = parse_args()

    notes = [read_dlopen_notes(filename) for filename in args.filenames]

    if args.raw:
        for filename, note in zip(args.filenames, notes):
            print(f'# {filename}')
            print_json(json.dumps(note, indent=2))

    if args.features is not None:
        features = group_by_feature(args.filenames, notes)

        toprint = {name:feature for name,feature in features.items()
                   if name in args.features or not args.features}
        if len(toprint) < len(args.features):
            sys.exit('Some features were not found')

        print('# grouped by feature')
        print_json(json.dumps(toprint,
                              indent=2,
                              default=lambda prio: prio.name))

    if args.sonames:
        sonames = group_by_soname(notes)
        for soname in sorted(sonames.keys()):
            print(f"{soname} {sonames[soname]}")
